File: backend/app/models/raw_collection.py
from datetime import datetime
from typing import Optional

from sqlmodel import Field, SQLModel


class RawCollection(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)

    item_name: str = Field(index=True)
    category: Optional[str] = Field(default=None, index=True)
    brand: Optional[str] = None

    shop_name: str = Field(index=True)
    location: Optional[str] = Field(default=None, index=True)

    price: float = Field(gt=0)
    quantity: float = Field(default=1, gt=0)
    unit: str = Field(default="unit")

    source: Optional[str] = None
    notes: Optional[str] = None

    status: str = Field(default="pending", index=True)
    collected_at: datetime = Field(default_factory=datetime.utcnow)
    reviewed_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

File: frontend/src/types/rawCollection.ts
import type { Item } from "./item";
import type { PriceObservation } from "./price";

export type RawCollectionStatus = "pending" | "approved" | "rejected";

export type RawCollection = {
  id: number;
  item_name: string;
  category: string | null;
  brand: string | null;
  shop_name: string;
  location: string | null;
  price: number;
  quantity: number;
  unit: string;
  source: string | null;
  notes: string | null;
  status: RawCollectionStatus;
  collected_at: string;
  reviewed_at: string | null;
  created_at: string;
};

export type CreateRawCollectionPayload = {
  item_name: string;
  category?: string | null;
  brand?: string | null;
  shop_name: string;
  location?: string | null;
  price: number;
  quantity: number;
  unit: string;
  source?: string | null;
  notes?: string | null;
  collected_at?: string;
};

export type UpdateRawCollectionPayload = Partial<CreateRawCollectionPayload>;

export type ApproveRawCollectionResult = {
  raw_collection: RawCollection;
  item: Item;
  price_observation: PriceObservation;
};

export type BulkApproveRawCollectionsResult = {
  approved_count: number;
  created_items_count: number;
  created_price_observations_count: number;
};

export type BulkRejectRawCollectionsResult = {
  rejected_count: number;
};
